Fix: Enable <18 Decimal Tokens as Collateral in Bonding Curve

0xNuggan opened the pull request:
changes to core contracts:
- padding the virtual collateral to have the same amount of decimals than the issued token.

main changes in tests:
- bound the fuzzed input in a way that considers the decimals the token has so we avoid overflows when padding
- use the padded version when we are directly calling the formula

0xNuggan committed:
change decimals to 6

0xNuggan committed:
working draft

some test are still failing because of excessive rejections

marvinkruse commented on src/modules/fundingManager/bondingCurveFundingManager/BancorVirtualSupplyBondingCurveFundingManager.sol:
```diff
@@ -454,7 +460,7 @@ contract BancorVirtualSupplyBondingCurveFundingManager is
         // issucance price, utilizes PPM for its computations. This leads to a precision loss
         // that's too significant to be acceptable for tokens with fewer than 7 decimals.
-        if (_decimals < 7) {
+        if (_decimals < 7 || _decimals > collateralTokenDecimals) {
```
Is this gonna be a problem for Tokens like USDC and Tether, as they are both 6 decimal tokens?

0xNuggan replied:
You are of course completely right, this is what I get for rushing to quickly update the setter before the dev call :laughing: It should be as mentioned in the interface. I pushed the fix

marvinkruse replied:
Haha, all good! 😂 
So the bonding curve (issuance) token decimals need to be at least 7 (for precision) and also need to be at least equal to the collateral tokens decimals, not lower, correct? That's how I understand the fix you pushed now, which looks good to me, just making sure.

Initially, I thought `_decimals` was for the collateral token, but it's for the issuance token, I think I get it now. Is there a good reason why someone would set the issuance tokens decimals to anything but 18? (just out of curiosity)

0xNuggan replied:
Yes, exactly! 
Personally, I can't think of a good reason to not have 18 decimals. But if we are going to "plug in" our bonding curve to an external token maybe that token had a good reason for it? So better to keep it open...

0xNuggan committed:
formatting

marvinkruse commented on test/modules/fundingManager/RebasingFundingManager.t.sol:
```diff
@@ -46,7 +46,7 @@ contract RebasingFundingManagerTest is ModuleTest {
 
     // Other constants.
-    uint8 private constant DECIMALS = 18;
+    //uint8 private constant DECIMALS = 18;  //Question : Is there any reason to hardcode the decimals for this fundingmanager to 18?
```
Is this for the collateral token or the issuance token? If it's for the collateral token, then it shouldn't be hardcoded, right?

edit: this might be for the issuance token as well, I think I initially misunderstood.

0xNuggan replied:
It's for the token that the bonding curve will issue. But I guess since we are going to make changes anyway for minting from an existing token, we shouldn't hardcode it

0xNuggan committed:
bugfix

0xNuggan committed:
fix vm.assume rejections

Zitzak commented on src/modules/fundingManager/bondingCurveFundingManager/BancorVirtualSupplyBondingCurveFundingManager.sol:
```diff
@@ -475,7 +478,12 @@ contract BancorVirtualSupplyBondingCurveFundingManager is
-        _addVirtualCollateralAmount(_depositAmount - feeAmount);
+        uint normalizedDepositAmount = _convertAmountToRequiredDecimal(
+            (_depositAmount - feeAmount),
+            collateralTokenDecimals,
```
As the amounts now get normalized to 18 decimals, if we indeed use for example USDC with 6 decimals as collateral, and a UI would like to get the virtual supply, don't they get this now in 18 decimals. So the dAPP interfacing with the contract has to know to calculate the value back to it's orginal decimals like this right?

Maybe it makes more sense to always convert it back to the original decimals, i.e. in my example this would be 6. Wdyt @0xNuggan

Zitzak commented on test/e2e/fundingManager/BondingCurveFundingManagerE2E.sol:
```diff
@@ -177,7 +177,8 @@ contract BondingCurveFundingManagerE2E is E2ETest {
             fundingManager.sell(
-                fundingManager.balanceOf(bob), fundingManager.balanceOf(bob)
+                fundingManager.balanceOf(bob),
+                fundingManager.balanceOf(bob) // Question: What happens with the fee? Shouldn't the amounts be wildly different since they are different tokens?
```
Unsure what is meant by the questions. Could you elaborate pls.

Zitzak commented on test/modules/fundingManager/bondingCurveFundingManager/BancorVirtualSupplyBondingCurveFundingManager.t.sol:
```diff
@@ -1467,12 +1669,22 @@ contract BancorVirtualSupplyBondingCurveFundingManagerTest is ModuleTest {
+        supply/collateral of 1/1, the first buy will move the price by a lot, since we will be effectively depositing 1e12 collateral tokens (1 + padding to 18). 
+        This will exceed the delta by a lot, but it IS expected behavior.
+        If we set the initial supply to 1e18 and the initial collateral t 1e12 (what we can call "realistic values"), the delta holds again. But on the ohter hand, we
+        aren't testing on the empty curve, just a "normal" point in it. 
```
Personally I would go with testing for realistic values. The BCRG is also building a simulation tool which will also (I think) help with defining the edge cases which we should then warn the user for

Zitzak reviewed (commented):
Only thing I found is the questions around always normalizing it, and how we should handle it for when the virtual supply is retrieved.

0xNuggan committed:
review fixes

modified the getVirtual... functions to return the value in the native decimal amount instead of the normalized one and added some tests

0xNuggan committed:
add normalization to virtualSupply

this should help set stuff up for external BC tokens

0xNuggan committed:
changes after discussion with marvin + corresponding tests

Ended up moving the place where the decimals get converted, and adapted all tests

0xNuggan committed:
remove remaining question

0xNuggan committed:
Merge branch 'temp_dev' into weird-erc-tokens-test

0xNuggan committed:
format

the online editor merge messed the format up

0xNuggan committed:
remove stray comments

Zitzak reviewed (approved):
LGTM 🚀

marvinkruse reviewed (approved):
lgtm as well! @0xNuggan
